controller/object_store_controller.go

```diff
@@ -0,0 +1,1190 @@
+	if err != nil {
+		return errors.Wrap(err, "API error while creating service")
+	}
+
+	endpoints, store, err := osc.getOrCreateS3Endpoints(store)
+	if err != nil {
+		return errors.Wrap(err, "API error while creating S3 ingresses")
+	}
+	osc.logger.Infof("object store %v has  %v S3 endpoint(s)", store.Name, len(endpoints))
+	// if there are no public endpoints, add the implicit cluster-internal one
+	if len(store.Status.Endpoints) == 0 {
+		store.Status.Endpoints = append(store.Status.Endpoints, fmt.Sprintf("%v.%v.svc", store.Name, osc.namespace))
+	}
+
+	if err := osc.checkVolume(vol); err != nil {
+		return nil
+	}
+
+	if err := osc.checkPV(pv); err != nil {
+		return nil
+	}
+
+	if err := osc.checkPVC(pvc); err != nil {
+		return nil
+	}
+
+	if err := osc.checkDeployment(dpl, store); err != nil {
+		return nil
+	}
+
+	logrus.Infof("object store %v is now running", store.Name)
+	store.Status.State = longhorn.ObjectStoreStateRunning
+	return nil
+}
+
+// This function does a short sanity check on the various resources that are
+// needed to operate the object stores. If any of them is found to be
+// unhealthy, the controller will transition the object store to "Error"
+// state, otherwise do nothing.
+func (osc *ObjectStoreController) handleRunning(store *longhorn.ObjectStore) (err error) {
+	if store.Spec.TargetState == longhorn.ObjectStoreStateStopped {
+		logrus.Infof("object store %v is now stopping", store.Name)
+		store.Status.State = longhorn.ObjectStoreStateStopping
+		return nil
+	}
+
+	dpl, err := osc.ds.GetDeployment(store.Name)
+	if err != nil && datastore.ErrorIsNotFound(err) {
+		return errors.Wrapf(err, "failed to find deployment %v", store.Name)
+	} else if err = osc.checkDeployment(dpl, store); err != nil {
+		logrus.Errorf("Object Store running but deployment not ready")
+		return err
+	}
+
+	_, err = osc.ds.GetService(osc.namespace, store.Name)
+	if err != nil && datastore.ErrorIsNotFound(err) {
+		return errors.Wrapf(err, "failed to find service %v", store.Name)
+	}
+
+	vol, err := osc.ds.GetVolume(genPVName(store))
+	if err != nil && datastore.ErrorIsNotFound(err) {
+		return errors.Wrapf(err, "failed to find volume %v", genPVName(store))
+	} else if err = osc.checkVolume(vol); err != nil {
+		logrus.Errorf("Object Store running but Volume not ready")
+		return err
+	}
+
+	pv, err := osc.ds.GetPersistentVolume(genPVName(store))
+	if err != nil && datastore.ErrorIsNotFound(err) {
+		return errors.Wrapf(err, "failed to find PV %v", genPVName(store))
+	} else if err = osc.checkPV(pv); err != nil {
+		logrus.Errorf("Object Store running but PV not ready")
+		return err
+	}
+
+	pvc, err := osc.ds.GetPersistentVolumeClaim(osc.namespace, genPVCName(store))
+	if err != nil && datastore.ErrorIsNotFound(err) {
+		return errors.Wrapf(err, "failed to find pvc %v", genPVCName(store))
+	} else if err = osc.checkPVC(pvc); err != nil {
+		logrus.Errorf("Object Store running but PVC not bound")
+		return err
+	}
+
+	return nil
+}
```

Comment:
To avoid potential nil panic and allow the ObjectStore to restart when there are missing dependent resources. WDYT about this:
```suggestion
func (osc *ObjectStoreController) handleRunning(store *longhorn.ObjectStore) (err error) {
	if store.Spec.TargetState == longhorn.ObjectStoreStateStopped {
		logrus.Infof("object store %v is now stopping", store.Name)
		store.Status.State = longhorn.ObjectStoreStateStopping
		return nil
	}

	dpl, err := osc.ds.GetDeployment(store.Name)
	if err != nil {
		if datastore.ErrorIsNotFound(err) {
			store.Status.State = longhorn.ObjectStoreStateError
		}
		return errors.Wrapf(err, "failed to find deployment %v", store.Name)
	} 
	if err = osc.checkDeployment(dpl, store); err != nil {
		logrus.Errorf("Object Store running but deployment not ready")
		return err
	}

	_, err = osc.ds.GetService(osc.namespace, store.Name)
	if err != nil {
		if datastore.ErrorIsNotFound(err) {
			store.Status.State = longhorn.ObjectStoreStateError
		}
		return errors.Wrapf(err, "failed to find service %v", store.Name)
	}

	vol, err := osc.ds.GetVolume(genPVName(store))
	if err != nil {
		if datastore.ErrorIsNotFound(err) {
			store.Status.State = longhorn.ObjectStoreStateError
		}
		return errors.Wrapf(err, "failed to find volume %v", genPVName(store))
	} 
	if err = osc.checkVolume(vol); err != nil {
		logrus.Errorf("Object Store running but Volume not ready")
		return err
	}

	pv, err := osc.ds.GetPersistentVolume(genPVName(store))
	if err != nil {
		if datastore.ErrorIsNotFound(err) {
			store.Status.State = longhorn.ObjectStoreStateError
		}
		return errors.Wrapf(err, "failed to find PV %v", genPVName(store))
	} 
	if err = osc.checkPV(pv); err != nil {
		logrus.Errorf("Object Store running but PV not ready")
		return err
	}

	pvc, err := osc.ds.GetPersistentVolumeClaim(osc.namespace, genPVCName(store))
	if err != nil {
		if datastore.ErrorIsNotFound(err) {
			store.Status.State = longhorn.ObjectStoreStateError
		}
		return errors.Wrapf(err, "failed to find pvc %v", genPVCName(store))
	} 
	if err = osc.checkPVC(pvc); err != nil {
		logrus.Errorf("Object Store running but PVC not bound")
		return err
	}

	return nil
}
```

Reply:
You'll have to explain to me why you think there could be a nil panic.
The [Go language spec](https://go.dev/ref/spec#Logical_operators) states that when multiple conditions are combined with operators like `&&`, the right hand conditional expression is evaluated only if necessary.
e.g. when `err == nil` and you have an expression like
```
if err != nil && foobar(err) {
}
```
the `foobar(err)` is never evaluated since the left hand conditional expression is already false and therefore the entire conditional expression is false regardless of the result of the right hand expression.

Reply:
Surely. Sorry, that didn't explain clearly enough.

Let's look at this code block:
```go
1	dpl, err := osc.ds.GetDeployment(store.Name)
2	if err != nil && datastore.ErrorIsNotFound(err) {
3		return errors.Wrapf(err, "failed to find deployment %v", store.Name)
4	} else if err = osc.checkDeployment(dpl, store); err != nil { 
5		logrus.Errorf("Object Store running but deployment not ready")
6		return err
7	}
```

Let say that we hit a temporary network errpr on line 1. So `err != nil` and `err` is not `ErrorIsNotFound`. At this moment, `dpl` is nil. Now we enter the line 4 and call function `osc.checkDeployment(dpl, store)` with a nil value for `dpl`. 
Inside the checkDeployment() function, the first thing is does is dereferencing the pointer `dpl` which mean it is dereferencing a nil pointer thus trigger a panic and crash Longhorn manager 
```go
func (osc *ObjectStoreController) checkDeployment(deployment *appsv1.Deployment, store *longhorn.ObjectStore) error {
	if *deployment.Spec.Replicas != 1 {
		deployment.Spec.Replicas = int32Ptr(1)
		osc.ds.UpdateDeployment(deployment)
```

Reply:
:facepalm: you're right. This would be fatal.

Reply:
I think we should set the objectstore to error so that we can restart it so recommending doing this:
```go
	dpl, err := osc.ds.GetDeployment(store.Name)
	if err != nil {
		if datastore.ErrorIsNotFound(err) {
			store.Status.State = longhorn.ObjectStoreStateError
		}
		return errors.Wrapf(err, "failed to find deployment %v", store.Name)
	} 
	if err = osc.checkDeployment(dpl, store); err != nil {
		logrus.Errorf("Object Store running but deployment not ready")
		return err
	}
```

Instead of the latest version:
```go
	dpl, err := osc.ds.GetDeployment(store.Name)
	if err != nil {
		if datastore.ErrorIsNotFound(err) {
			return errors.Wrapf(err, "failed to find deployment %v", store.Name)
		}
	} else if err = osc.checkDeployment(dpl, store); err != nil {
		logrus.Errorf("Object Store running but deployment not ready")
		return err
	}
```

WDYT?

Reply:
The `reconcile` function will update the object store's state to error if any of the `handleRunning`, `handleStarting`, `handleStopping`, `handleStopped` or `initializeObjectStore` functions return an error. Therefore it's sufficient to return an error here, the object store's state will be updated accordingly.